presto-hudi/src/main/java/com/facebook/presto/hudi/HudiFileSkippingManager.java

```diff
@@ -0,0 +1,295 @@
+            log.warn(e, "failed to do data skipping for table: %s, fallback to all files scan", metaClient.getBasePathV2());
+            candidateFileSlices = allInputFileSlices;
+        }
+        if (log.isDebugEnabled()) {
+            int candidateFileSize = candidateFileSlices.values().stream().mapToInt(List::size).sum();
+            int totalFiles = allInputFileSlices.values().stream().mapToInt(List::size).sum();
+            double skippingPercent = totalFiles == 0 ? 0.0d : (totalFiles - candidateFileSize) / (totalFiles + 0.0d);
+            log.debug("Total files: %s; candidate files after data skipping: %s; skipping percent %s", totalFiles, candidateFileSize, skippingPercent);
+        }
+        return candidateFileSlices;
+    }
+
+    private Map<String, List<FileSlice>> lookupCandidateFilesInMetadataTable(Map<String, List<FileSlice>> inputFileSlices, TupleDomain<ColumnHandle> tupleDomain)
+    {
+        // split regular column predicates
+        TupleDomain<HudiColumnHandle> regularTupleDomain = HudiPredicates.from(tupleDomain).getRegularColumnPredicates();
+        TupleDomain<String> regularColumnPredicates = regularTupleDomain.transform(HudiColumnHandle::getName);
+        if (regularColumnPredicates.isAll() || !regularColumnPredicates.getDomains().isPresent()) {
+            return inputFileSlices;
+        }
+        List<String> regularColumns = regularColumnPredicates.getDomains().get().entrySet().stream().map(Map.Entry::getKey).collect(Collectors.toList());
+        // get filter columns
+        List<String> encodedTargetColumnNames = regularColumns.stream().map(col -> new ColumnIndexID(col).asBase64EncodedString()).collect(Collectors.toList());
+        Map<String, List<HoodieMetadataColumnStats>> statsByFileName = metadataTable.getRecordsByKeyPrefixes(
+                encodedTargetColumnNames,
+                HoodieTableMetadataUtil.PARTITION_NAME_COLUMN_STATS, true)
+                .collectAsList()
+                .stream()
+                .filter(f -> f.getData().getColumnStatMetadata()
+                        .isPresent())
+                .map(f -> f.getData().getColumnStatMetadata().get()).collect(Collectors.groupingBy(HoodieMetadataColumnStats::getFileName));
+
+        // prune files.
+        return inputFileSlices.entrySet().parallelStream().collect(Collectors.toMap(entry -> entry.getKey(), entry -> {
+            return entry.getValue().stream().filter(fileSlice -> {
+                String fileSliceName = fileSlice.getBaseFile().map(BaseFile::getFileName).orElse("");
+                // no stats found
+                if (!statsByFileName.containsKey(fileSliceName)) {
+                    return true;
+                }
+                List<HoodieMetadataColumnStats> stats = statsByFileName.get(fileSliceName);
+                return evaluateStatisticPredicate(regularColumnPredicates, stats, regularColumns);
+            }).collect(Collectors.toList());
```

Comment:
it would be better to extract a method for this block to get better readability.